target/linux/ath79/generic/base-files/lib/upgrade/openmesh.sh

```diff
@@ -0,0 +1,114 @@
+		do
+			[ "$cfg_opt" = "[$section]" ] && our_section=1 && continue
+			[ "$our_section" = "1" ] || continue
+
+			our_param=$(echo ${cfg_opt%%=*})
+			[ "$param" = "$our_param" ] && echo ${cfg_opt##*=} && break
+		done
+}
+
+platform_do_upgrade_openmesh()
+{
+	local img_path="$1"
+	local restore_backup
+
+	local setenv_script="/tmp/fw_env_upgrade"
+
+	local inactive_mtd="$(find_mtd_index $PART_NAME)"
+	local inactive_offset="$(cat /sys/class/mtd/mtd${inactive_mtd}/offset)"
+	local total_size="$(cat /sys/class/mtd/mtd${inactive_mtd}/size)"
+	local flash_start_mem=0x9f000000
+	local data_offset=$((64 * 1024))
+
+	# detect to which flash region the new image is written to.
+	#
+	# 1. check what is the mtd index for the first flash region on this
+	#    device
+	# 2. check if the target partition ("inactive") has the mtd index of
+	#    the first flash region
+	#
+	#    - when it is: the new bootseq will be 1,2 and the first region is
+	#      modified
+	#    - when it isnt: bootseq will be 2,1 and the second region is
+	#      modified
+	#
+	# The detection has to be done via the hardcoded mtd partition because
+	# the current boot might be done with the fallback region. Let us
+	# assume that the current bootseq is 1,2. The bootloader detected that
+	# the image in flash region 1 is corrupt and thus switches to flash
+	# region 2. The bootseq in the u-boot-env is now still the same and
+	# the sysupgrade code can now only rely on the actual mtd indexes and
+	# not the bootseq variable to detect the currently booted flash
+	# region/image.
+	#
+	# In the above example, an implementation which uses bootseq ("1,2") to
+	# detect the currently booted image would assume that region 1 is booted
+	# and then overwrite the variables for the wrong flash region (aka the
+	# one which isn't modified). This could result in a device which doesn't
+	# boot anymore to Linux until it was reflashed with ap51-flash.
+	local next_boot_part="1"
+	local primary_kernel_mtd="3"
+	[ "$inactive_mtd" = "$primary_kernel_mtd" ] || next_boot_part="2"
+
+	local cfg_size=$(dd if="$img_path" bs=2 skip=35 count=4 2>/dev/null)
+	local cfg_content=$(dd if="$img_path" bs=1 skip=$data_offset count=$((0x$cfg_size)) 2>/dev/null)
+
+	local kernel_size=$(dd if="$img_path" bs=2 skip=71 count=4 2>/dev/null)
+	local kernel_length=$((0x$kernel_size))
+	local kernel_md5=$(cfg_value_get "$cfg_content" "vmlinux" "md5sum")
+
+	local rootfs_size=$(dd if="$img_path" bs=2 skip=107 count=4 2>/dev/null)
+	local rootfs_md5=$(cfg_value_get "$cfg_content" "rootfs" "md5sum")
+	local rootfs_checksize=$(cfg_value_get "$cfg_content" "rootfs" "checksize")
+
+	# take care of restoring a saved config
+	[ -n "$UPGRADE_BACKUP" ] && restore_backup="${MTD_CONFIG_ARGS} -j ${UPGRADE_BACKUP}"
+
+	# write image parts
+	mtd -q erase inactive
+	dd if="$img_path" bs=1 skip=$(($data_offset + 0x$cfg_size + 0x$kernel_size)) count=$((0x$rootfs_size)) 2>&- | mtd -n -p $kernel_length $restore_backup write - $PART_NAME
```

Comment:
bs=1 will be _very_ slow here? One should consider to improve that. But probably in a separate patch ...

Reply:
Yes, it is rather slow. Increases the write time of the default image (without feeds) on an OM2P-HS v3 from roughly 30s to 120s. I will continue the discussion in the other comment